c/tskit/trees.c

```diff
@@ -7644,3 +7644,417 @@ tsk_treeseq_divergence_matrix(const tsk_treeseq_t *self, tsk_size_t num_samples,
+    }
+    ret = tsk_edge_table_keep_rows(edges, keep, 0, NULL);
+out:
+    tsk_blkalloc_free(&edge_list_heap);
+    tsk_tree_position_free(&tree_pos);
+    tsk_safe_free(degree);
+    tsk_safe_free(out_parent);
+    tsk_safe_free(keep);
+    tsk_safe_free(not_sample);
+    return ret;
+}
+
+static int
+tsk_treeseq_slide_mutation_nodes_up(
+    const tsk_treeseq_t *self, tsk_mutation_table_t *mutations)
+{
+    int ret = 0;
+    bool valid;
+    double t, right;
+    tsk_id_t c, p, tj, e, next_mut;
+    tsk_tree_position_t tree_pos;
+    const tsk_table_collection_t *tables = self->tables;
+    const tsk_size_t num_nodes = tables->nodes.num_rows;
+    const tsk_edge_table_t *edges = &tables->edges;
+    tsk_id_t *parent = tsk_malloc(num_nodes * sizeof(*parent));
+    double *sites_position = tables->sites.position;
+    double *nodes_time = tables->nodes.time;
+
+    tsk_memset(&tree_pos, 0, sizeof(tree_pos));
+
+    if (parent == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(parent, 0xff, num_nodes * sizeof(*parent));
+
+    ret = tsk_tree_position_init(&tree_pos, self, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    valid = tsk_tree_position_next(&tree_pos);
+    next_mut = 0;
+
+    while (valid) {
+        right = tree_pos.interval.right;
+
+        for (tj = tree_pos.out.start; tj != tree_pos.out.stop; tj += TSK_DIR_FORWARD) {
+            e = tree_pos.out.order[tj];
+            parent[edges->child[e]] = TSK_NULL;
+        }
+        for (tj = tree_pos.in.start; tj != tree_pos.in.stop; tj += TSK_DIR_FORWARD) {
+            e = tree_pos.in.order[tj];
+            parent[edges->child[e]] = edges->parent[e];
+        }
+        while (next_mut < (tsk_id_t) mutations->num_rows
+               && sites_position[mutations->site[next_mut]] < right) {
+            t = mutations->time[next_mut];
+            if (tsk_is_unknown_time(t)) {
+                ret = TSK_ERR_DISALLOWED_UNKNOWN_MUTATION_TIME;
+                goto out;
+            }
+            c = mutations->node[next_mut];
+            p = parent[c];
+            while (p != TSK_NULL && nodes_time[p] <= t) {
+                c = p;
+                p = parent[c];
+            }
+            tsk_bug_assert(nodes_time[c] <= t);
+            mutations->node[next_mut] = c;
+            next_mut++;
+        }
+
+        valid = tsk_tree_position_next(&tree_pos);
+    }
+
+out:
+    tsk_safe_free(parent);
+
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
+    tsk_flags_t TSK_UNUSED(options), tsk_treeseq_t *output)
+{
+    int ret = 0;
+    tsk_table_collection_t tables;
+    tsk_treeseq_t ts;
+    int iter, j;
+    tsk_size_t last_num_edges;
+    const int direction[] = { TSK_DIR_FORWARD, TSK_DIR_REVERSE };
+
+    tsk_memset(&tables, 0, sizeof(tables));
+    tsk_memset(&ts, 0, sizeof(ts));
+    tsk_memset(output, 0, sizeof(*output));
+
```

Comment:
Probably good to check that max_iter > 0. It's not helpful to silently do nothing here I think? Negative max_iter is surely an error anyway.